### ISSUE TITLE
`needless_pass_by_value` lint does not trigger on `async fn`

### ISSUE THREAD

**Issue by skr2005** [rust-lang/rust-clippy#15136]
### Summary

The `needless_pass_by_value` lint does not fire for `async fn`, which makes it ineffective in projects that use async functions extensively. I am not sure whether this behavior is intended though.

### Lint Name

needless_pass_by_value

### Reproducer

I tried this code:

```rust
#![forbid(clippy::needless_pass_by_value)]

use std::hint::black_box;

fn main() {
    black_box(take_value_1);
    black_box(take_value_2);
}

struct Foo;

async fn take_value_1(x: Foo) {
    black_box(&x);
}

fn take_value_2(x: Foo) {
    black_box(&x);
}
```

I expected clippy to emit an error for both `take_value_1` and `take_value_2`, since neither function consumes its `Foo` argument.

However, only `take_value_2` triggers the lint. `take_value_1` does not:
```
error: this argument is passed by value, but not consumed in the function body
  --> src\main.rs:16:20
   |
16 | fn take_value_2(x: Foo) {
   |                    ^^^
   |
help: or consider marking this type as `Copy`
  --> src\main.rs:10:1
   |
10 | struct Foo;
   | ^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_value
note: the lint level is defined here
  --> src\main.rs:1:11
   |
1  | #![forbid(clippy::needless_pass_by_value)]
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider taking a reference instead
   |
16 | fn take_value_2(x: &Foo) {
   |                    +

error: could not compile `tt` (bin "tt") due to 1 previous error
```

### Version

```text
rustc 1.87.0 (17067e9ac 2025-05-09)
binary: rustc
commit-hash: 17067e9ac6d7ecb70e50f92c1944e545188d2359
commit-date: 2025-05-09
host: x86_64-pc-windows-gnu
release: 1.87.0
LLVM version: 20.1.1
```

**Comment by samueltardieu** [comment]
Using a reference in `async fn take_value_1(x: &Foo)` would restrict the lifetime of the returned anonymous future to the one of `x`, as its value would not be moved into the future. For example the following code would no longer compile:

```rust
use std::hint::black_box;

fn run_future(f: impl std::future::Future<Output = ()> + 'static) {
    black_box(f);
}

fn main() {
    let x = Foo;
    run_future(take_value_1(x));
}

struct Foo;

async fn take_value_1(x: Foo) {
    black_box(x);
}
```

So unless we know for sure that the future will never be passed to something that require a lifetime greater than the one of its argument, we cannot suggest using a reference there.

**Comment by skr2005** [comment]
Yes. Now I understand. It seems non-trivial for clippy to check lifetimes in which the futures are used, so feel free to close this issue as not planned.

**Comment by sivizius** [comment]
This was somehow fixed in 1.88.0, see https://github.com/rust-lang/rust-clippy/issues/14417#issuecomment-3163679489 and https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=6963e3799c8da8f6af3423d929cb90c1. Can you confirm, @skr2005?

**Comment by skr2005** [comment]
@sivizius  I tried your playground link, and clippy doesn't report error on `take_value_1`. I also tried Rust 1.88 on my machine, and clippy doesn't report error on `take_value_1`, either. Maybe the problem is not fixed, or am I missing something to make it work?

**Comment by sivizius** [comment]
Ah, sorry, I misunderstood the problem m). I just investigated the linked issue regarding cognitive-complexity, after updating rustc which revealed these previous false-negatives and thought, this fixed this issue too. I didn’t interpreted the error on the playground correctly.
